localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_batch.py

```diff
@@ -0,0 +1,202 @@
+        self, env: Environment, resource_runtime_part: ResourceRuntimePart, raw_parameters: dict
+    ) -> None:
+        if self.resource.condition == ResourceCondition.Sync:
+            self._attach_aws_environment_variables(parameters=raw_parameters)
+        super()._before_eval_execution(
+            env=env, resource_runtime_part=resource_runtime_part, raw_parameters=raw_parameters
+        )
+
+    def _from_error(self, env: Environment, ex: Exception) -> FailureEvent:
+        if isinstance(ex, ClientError):
+            error_code = ex.response["Error"]["Code"]
+            error_name = f"Batch.{error_code}"
+            status_code = ex.response["ResponseMetadata"]["HTTPStatusCode"]
+            error_message = ex.response["Error"]["Message"]
+            request_id = ex.response["ResponseMetadata"]["RequestId"]
+            response_details = "; ".join(
+                [
+                    "Service: AWSBatch",
+                    f"Status Code: {status_code}",
+                    f"Error Code: {error_code}",
+                    f"Request ID: {request_id}",
+                    "Proxy: null",
+                ]
+            )
+            cause = f"Error executing request, Exception : {error_message}, RequestId: {request_id} ({response_details})"
+            return FailureEvent(
+                env=env,
+                error_name=CustomErrorName(error_name),
+                event_type=HistoryEventType.TaskFailed,
+                event_details=EventDetails(
+                    taskFailedEventDetails=TaskFailedEventDetails(
+                        error=error_name,
+                        cause=cause,
+                        resource=self._get_sfn_resource(),
+                        resourceType=self._get_sfn_resource_type(),
+                    )
+                ),
+            )
+        return super()._from_error(env=env, ex=ex)
+
+    def _eval_service_task(
+        self,
+        env: Environment,
+        resource_runtime_part: ResourceRuntimePart,
+        normalised_parameters: dict,
+    ):
+        service_name = self._get_boto_service_name()
+        api_action = self._get_boto_service_action()
+        batch_client = boto_client_for(
+            region=resource_runtime_part.region,
+            account=resource_runtime_part.account,
+            service=service_name,
+        )
+        response = getattr(batch_client, api_action)(**normalised_parameters)
+        response.pop("ResponseMetadata", None)
+        env.stack.append(response)
+
+    def _sync_to_submit_job(
+        self,
+        env: Environment,
+        resource_runtime_part: ResourceRuntimePart,
+    ) -> None:
+        batch_client = boto_client_for(
+            region=resource_runtime_part.region,
+            account=resource_runtime_part.account,
+            service="batch",
+        )
+        submission_output: dict = env.stack.pop()
+        job_id = submission_output["JobId"]
+
+        def _has_terminated() -> Optional[dict]:
```

Comment:
typing nit: we could (at some point) consider migrating to `dict | None` at some point or for new code (now natively supported)